plug-ins/prometheus/architecture.md

```diff
@@ -0,0 +1,127 @@
+plugin will have it's own particular update cycle based on what makes sense for the particular plugin.  During each update cycle, the instrumented plugins will post the collection of metrics to the prometheus_plugin (if enabled).  The prometheus_plugin will
+update its internal data model as these messages are received.
+
+## Prometheus Endpoint
+The prometheus_plugin will have one endpoint (/metrics) to support the pull functionality. The pull endpoint will be implemented on top of nodeos http_plugin, rather than civetweb.
+Data will be exported using the prometheus text-based exposition format.
+The http endpoint for the prometheus_plugin will utilize the prometheus strand to send client updates.  The plugin will receive requests on an http thread, which will immediately post a lambda to the prometheus strand.  The request will pull the
+updated metric data from the prometheus data model and use the prometheus-cpp library to perform the serialization.  Requests will be bound by a timeout of 30ms.
+
+## Instrumentation of plugins
+
+Each of the instrumented plugins will have a register_metrics_listener() method to register a std::function that takes a vector of runtime_metrics.  Instrumented plugins will each maintain a specialized plugin_metrics structure (see below).  The plugin will pass the register_listener() method to the plugin_metrics struct. The plugin_metrics struct will manage the update cycle
+to enforce a clean 'snapshot' view of all the metric data for a plugin.
+
+A plugin_metrics base class will be provided for plugins to support collecting/publishing metrics to the prometheus_plugin (or potentially some other metrics publisher in the future). Each plugin will extend the base class and include runtime_metrics as members of the base class.
+Each plugin implements a virtual function metrics(), which will collect all relevant metrics and insert them into a vector for processing by the prometheus_plugin.
+
+The register_listener() method on plugin_metrics will be used by the prometheus plugin to register a lambda to perform the metrics update.
+The lambda will post to a strand, backed by a named_thread pool (consisting of a single thread).  The lambda posted to the strand will perform the metrics update inside the prometheus_plugin.
+Each plugin will have its own update cycle - for instance the producer_plugin will update on each block.  The instrumented plugins will report their metrics update via the post_metrics() method on the plugin_metrics object.
+The post_metrics method will check the should_post() method to verify that it should actually update the prometheus_plugin by calling should_post().  The should_post method will check to see if there is a registered listener and verify that enough time has elapsed since
+the last post.  The default throttling will be to ensure that there is no more than one post per 250ms.
```

Comment:
This interval seems needlessly broad as a minimum.  A default of 250ms is fine, but a minimum of 100ms makes for much more friendly (i.e. smoother, more granular) strip charts.  I've been using GKrellM for system monitoring at 10Hz for nearly 30 years, and that level of granularity is very useful for letting you see transient spikes while being less misleading about their size.

Populating the data structures described and posting them looks very light weight for the primary thread, so a slightly higher refresh rate shouldn't be an undue burden.